pkg/cableengine/syncgatewaystatus.go

```diff
@@ -0,0 +1,161 @@
+package cableengine
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+
+	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/log"
+)
+
+const GatewayUpdateIntervalSeconds = 5
+const GatewayStaleTimeoutSeconds = GatewayUpdateIntervalSeconds * 3
+const updateTimestampAnnotation = "update-timestamp"
+
+func (i *engine) syncGatewayStatus() {
+
+	i.Lock()
+	defer i.Unlock()
+
+	klog.V(log.TRACE).Info("Running syncGatewayStatus()")
+
+	gatewayObj, err := i.generateGatewayObject()
+	if err != nil {
+		klog.Errorf("generating gateway object from driver connections: %s", err)
+		return
+	}
+
+	existingGw, err := i.getLastSyncedGateway(gatewayObj.Name)
+
+	// log and stop for any error different to a not-found error
+	if err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("trying to read existing gateway from k8s api: %s", err)
+		return
+	}
+
+	gwClient := i.clientset.SubmarinerV1().Gateways(i.namespace)
+
+	if err != nil && errors.IsNotFound(err) {
+		klog.V(log.TRACE).Infof("Gateway object needs creation: %+v", gatewayObj)
+		_, err = gwClient.Create(gatewayObj)
```

Comment:
Does this patch depend on any other PR?

I tried this PR locally and see that following error continuously repeated.
 I0417 09:12:46.798166       1 syncgatewaystatus.go:46] Gateway object needs creation: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:cluster2-worker GenerateName: Namespace: SelfLink: U │
│ ID: ResourceVersion: Generation:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[ha-status:active] Annotations:map[up │
│ ate-timestamp:1587114766[] OwnerReferences:[] Initializers:nil Finalizers:[] ClusterName:} Status:{Version:dev HAStatus:ACTIVE LocalEndpoint:{ClusterID:cluster2 CableName:submariner-cabl │
│ e-cluster2-172-17-0-8 Hostname:cluster2-worker Subnets:[100.92.0.0/16 10.242.0.0/16] PrivateIP:172.17.0.8 PublicIP: NATEnabled:false Backend:strongswan BackendConfig:map[]} Connections:[ │
│ {Status:CONNECTED StatusMessage:Connected to 172.17.0.5:4500 - encryption alg=AES_GCM_16, keysize=128 rekey-time=12786 Endpoint:{ClusterID:cluster3 CableName:submariner-cable-cluster3-17 │
│ 2-17-0-5 Hostname:cluster3-worker Subnets:[100.93.0.0/16 10.243.0.0/16] PrivateIP:172.17.0.5 PublicIP: NATEnabled:false Backend:strongswan BackendConfig:map[]}}]}}                        │
│ E0417 09:12:46.798660       1 syncgatewaystatus.go:49] creating Gateway object: the server could not find the requested resource (post gateways.submariner.io)                             │
│ I0417 09:12:51.799029       1 syncgatewaystatus.go:27] Running syncGatewayStatus()                                                                                                         │
│ I0417 09:12:51.802367       1 syncgatewaystatus.go:159] generateGatewayObject: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:cluster2-worker GenerateName: Namespace: SelfLink: UID: Reso │
│ urceVersion: Generation:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[ha-status:active] Annotations:map[update-tim │
│ stamp:1587114771[] OwnerReferences:[] Initializers:nil Finalizers:[] ClusterName:} Status:{Version:dev HAStatus:ACTIVE LocalEndpoint:{ClusterID:cluster2 CableName:submariner-cable-cluste │
│ r2-172-17-0-8 Hostname:cluster2-worker Subnets:[100.92.0.0/16 10.242.0.0/16] PrivateIP:172.17.0.8 PublicIP: NATEnabled:false Backend:strongswan BackendConfig:map[]} Connections:[{Status: │
│ CONNECTED StatusMessage:Connected to 172.17.0.5:4500 - encryption alg=AES_GCM_16, keysize=128 rekey-time=12781 Endpoint:{ClusterID:cluster3 CableName:submariner-cable-cluster3-172-17-0-5 │
│  Hostname:cluster3-worker Subnets:[100.93.0.0/16 10.243.0.0/16] PrivateIP:172.17.0.5 PublicIP: NATEnabled:false Backend:strongswan BackendConfig:map[]}}]}}                                │
│ I0417 09:12:51.804919       1 syncgatewaystatus.go:123] getLastSyncedGateway: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name: GenerateName: Namespace: SelfLink: UID: ResourceVersion: Ge │
│ neration:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Initializers:nil Fin │
│ alizers:[] ClusterName:} Status:{Version: HAStatus: LocalEndpoint:{ClusterID: CableName: Hostname: Subnets:[] PrivateIP: PublicIP: NATEnabled:false Backend: BackendConfig:map[]} Connecti │
│ ons:[]}}

Reply:
You need to remove your local shipyard image so it will pull the latest one with the latest subctl which includes it.

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: gateways.submariner.io
spec:
  conversion:
    strategy: None
  group: submariner.io
  names:
    kind: Gateway
    listKind: GatewayList
    plural: gateways
    singular: gateway
  scope: Namespaced
  version: v1
  versions:
  - name: v1
    served: true
    storage: true

```

Reply:
You can apply that CRD manually otherwise :-)

Reply:
Thanks for the clarification. Yes, it works fine after updating the shipyard image.